keps/sig-network/20200127-disable-service-type-lb-nodeports.md

```diff
@@ -0,0 +1,167 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+As of today, a Kubernetes Service with Type=LoadBalancer will always allocate a node port for every service port. Though most implementations of Service Type=LoadBalancer do require node ports, there are several implementations that do not. This KEP proposes to add a new flag on the apiserver (`--disable-loadbalancer-node-ports`) to optionally disable node port allocation for Service Type=LoadBalancer.
+
+## Motivation
+
+There are several implementations of Service Type=LoadBalancer API today that do not consume the node ports automatically allocated by Kubernetes. Some examples include but are not limited to:
+* [MetalLB](https://github.com/danderson/metallb)
+* [kube-router](https://github.com/cloudnativelabs/kube-router)
+
+The drawbacks of allocating unused node ports are that:
+* the number of load balancers is now limited to the number of available node ports
+* node ports are exposed for an LB even though they are not used -- unecessarily exposed ports can fail regulatory/compliance requirements
+
+For clusters that have integrations for Service Type=LoadBalancer but don't require node ports should have the option to disable node port allocation.
+
+### Goals
+
+* disable node port for Service Type=LoadBalancer
+* don't change any existing default behavior for Service Type=LoadBalancer
+* Service LoadBalancer node ports can safely be re-enabled in a cluster
+
+## Proposal
+
+A flag `--disable-loadbalancer-node-ports` will be added to the apiserver to disable node port allocations for Service Type=LoadBalancer. The flag's default value is `false` to ensure existing behavior is preserved.
+
+If `--disable-loadbalancer-node-ports` is `true`, the following changes are expected on the apiserver:
+- any new Service LoadBalancers will not have node ports allocated
+- existing Service LoadBalancers with node ports will continue to operate with node ports
+- new Service LoadBalancers can still get node ports assigned manually
+
+The service port allocator controller will be used to rectify the following edge cases:
+- Service LoadBalancer node port allocation is disabled and then re-enabled -- any new Services that weren't allocated node ports should get node ports assigned by the port allocator controller.
+
+## Design Details
+
+### Global API Server Flag
+
+A global apiserver flag `--disable-loadbalancer-node-ports` will be used to toggle this behavior as opposed to per-Service annotations or webhooks since Service LoadBalancer implementations are also done at a global level (via the cloud provider).
+Since a cluster is very unlikely to have multiple implementations of Service LoadBalancers, it wouldn't make sense to allocate node ports for a subset of Service LoadBalancers.
```

Comment:
I don't think so, on metal you could easily have two types (either by protocol, like BGP vs some form of VRRP) or you could have mixed clusters without cloud provider that do custom integrations.  It's rare on clouds, but not rare on not-clouds.

Reply:
Is it reasonable to expect the cluster admins who configure the apiserver to know ALL of the LB implementations in use on that cluster?  If so then a flag on apiserver may be viable.

Alternatively, is it acceptable to force every user to set a flag on every service (or for the cluster admin to do this via an admission controller)?  I posit that it is NOT acceptable to make users do it, and that built-in admission control is not much different from a flag.  If we expect cluster admins to run their own admission controller, that may be viable.

The risk of a flag on apiserver is that, during upgrade, different servers might behave differently, should not really be a problem as long as other apiservers and controllers TRULY do not need the field.